app/src/main/java/com/kamron/pogoiv/MainActivity.java

```diff
@@ -145,12 +147,12 @@ protected void onCreate(Bundle savedInstanceState) {
         npTrainerLevel.setWrapSelectorWheel(false);
         npTrainerLevel.setValue(trainerLevel);
 
-        Button launch = (Button) findViewById(R.id.start);
-        launch.setOnClickListener(new View.OnClickListener() {
+        launchButton = (Button) findViewById(R.id.start);
+        launchButton.setOnClickListener(new View.OnClickListener() {
 
             @Override
             public void onClick(View v) {
-                if (((Button) v).getText().toString().equals(getString(R.string.main_permission))) {
+                if (!hasAllPermissions()) {
```

Comment:
Feel free to do that for the new code (I'm indifferent, and @sarav's was phrased as a slight preference). I don't think you have to.

But please don't move around the existing code, both because it's not part of this PR and to reduce conflicts with other changes.